src/Contacts/PhoneNumberUtil.m

```diff
@@ -32,13 +34,45 @@ - (instancetype)init {
     if (self) {
         _nbPhoneNumberUtil = [[NBPhoneNumberUtil alloc] init];
         _countryCodesFromCallingCodeCache = [NSMutableDictionary new];
+        _parsedPhoneNumberCache = [NSCache new];
 
         OWSSingletonAssert();
     }
 
     return self;
 }
 
+- (nullable NBPhoneNumber *)parse:(NSString *)numberToParse
+                    defaultRegion:(NSString *)defaultRegion
+                            error:(NSError **)error
+{
+    NSString *hashKey = [NSString stringWithFormat:@"numberToParse:%@defaultRegion:%@", numberToParse, defaultRegion];
+
+    NBPhoneNumber *result = [self.parsedPhoneNumberCache objectForKey:hashKey];
+
+    if (!result) {
+        result = [self.nbPhoneNumberUtil parse:numberToParse defaultRegion:defaultRegion error:error];
+        if (result) {
```

Comment:
Also `error` parameter will be nil the second time we try to parse an invalid phone number, which might affect logging at caller. NBD?